tests/utilities/test_distributed.py

```diff
@@ -42,3 +43,48 @@ def test_collect_states():
     """
     tutils.set_random_main_port()
     mp.spawn(_test_collect_states, args=(2,), nprocs=2)
+
+
+def _test_all_gather_uneven_tensors(rank, world_size):
+    os.environ["MASTER_ADDR"] = "localhost"
+
+    torch.cuda.set_device(f"cuda:{rank}")
+
+    # initialize the process group
+    torch.distributed.init_process_group("nccl", rank=rank, world_size=world_size)
+
+    tensor = torch.ones(rank)
+    result = gather_all_tensors(tensor)
+    assert len(result) == world_size
+    for idx in range(world_size):
+        assert len(result[idx]) == idx
+        assert (result[idx] == torch.ones_like(result[idx])).all()
+
+
+def _test_all_gather_uneven_tensors_multidim(rank, world_size):
+    os.environ["MASTER_ADDR"] = "localhost"
+
+    torch.cuda.set_device(f"cuda:{rank}")
+
+    # initialize the process group
+    torch.distributed.init_process_group("nccl", rank=rank, world_size=world_size)
+    tensor = torch.ones(rank + 1, 2 - rank)
+    result = gather_all_tensors(tensor)
+    assert len(result) == world_size
+    for idx in range(world_size):
+        val = result[idx]
+        assert val.shape == (idx + 1, 2 - idx)
+        assert (val == torch.ones_like(val)).all()
+
+
+@RunIf(min_gpus=2, min_torch="1.10", skip_windows=True)
```

Comment:
is this necessary? i.e., is any of the implementation dependent on GPU or would gloo on cpu work too?